### PR TITLE
chore(ux): improve error message when attaching without subject artifact 

### LINKED ISSUE #1424: improve error message of `oras attach` when subject artifact is not specified
When using `oras attach`, if a user forgot to provide the subject artifact reference and only specifies one to-be-uploaded file, ORAS CLI will output (tested in bash, different shell might have different output):

```console
> ARTIFACT=
> oras attach --artifact-type test/example $ARTIFACT./test.json
Error: neither file nor annotation provided in the command
Usage: oras attach [flags] --artifact-type=<type> <name>{:<tag>|@<digest>} <file>[:<layer_media_type>] [...]
To attach to an existing artifact, please provide files via argument or annotations via flag "--annotation". Run "oras attach -h" for more options and examplesdigest]
```

Input was parsed under a wrong consumption: user does specify subject artifact reference `./layers`, which actually is a file reference to be attached. This error can be avoided by parsing the only input argument. If the reference is valid, continue with current output. If it's not, give corresponding hint like:

```console
Error: "./test.json" is an invalid artifact reference
Did you forget to specify a subject artifact?
```

### PR DESCRIPTION
**Pull request by qweeah**
**What this PR does / why we need it**:
This PR improves error message when the `oras attach` is run with only one argument, which might be an invalid artifact subject reference or mistakenly used as a to-be-uploaded file path. 

**Before**
```console
> oras attach --artifact-type test/example $ARTIFACT ./layers 
Error: neither file nor annotation provided in the command
Usage: oras attach [flags] --artifact-type=<type> <name>{:<tag>|@<digest>} <file>[:<layer_media_type>] [...]
To attach to an existing artifact, please provide files via argument or annotations via flag "--annotation". Run "oras attach -h" for more options and examples
```
**After**
```console
> oras attach --artifact-type test/example $ARTIFACT ./layers 
Error: "./layers": no tag or digest specified
Usage: oras attach [flags] --artifact-type=<type> <name>{:<tag>|@<digest>} <file>[:<layer_media_type>] [...]
Have you specified an artifact reference to attach to? Please specify a reference in the form of "<name>:<tag>" or "<name>@<digest>". Run "oras attach -h" for more options and examples
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1424, #1404

**Please check the following list**:
- [x]  Does the affected code have corresponding tests, e.g. unit test, E2E test?
- [ ]  Does this change require a documentation update?
- [ ]  Does this introduce breaking changes that would require an announcement or bumping the major version?
- [ ]  Do all new files have an appropriate license header?

### CODE REVIEW

**Comment by wangxiaoxuan273** [comment]
LGTM but I'm not a maintainer

File: cmd/oras/internal/errors/errors.go [hunk 1]
```diff
@@ -26,6 +26,15 @@ import (
 	"oras.land/oras-go/v2/registry/remote/errcode"
 )
 
+// OperationType reperesnts the type of operation.
+type OperationType string
+
+const (
+	// ParseArtifactReference is the operation type parsing artifact
+	// reference.
+	ParseArtifactReference = OperationType("")
```
Comment (TerryHowe) [inline cmd/oras/internal/errors/errors.go:35]: Set this to `ParseArtifactReference` maybe?

Comment (qweeah) [reply cmd/oras/internal/errors/errors.go:35]: Sorry I don't get it.

Comment (qweeah) [reply cmd/oras/internal/errors/errors.go:35]: Do you mean change the string value to `"ParseArtifactReference "`

Comment (TerryHowe) [reply cmd/oras/internal/errors/errors.go:35]: Yes or maybe use iota instead?

**Review by TerryHowe — APPROVED** [review]
/lgtm

File: cmd/oras/root/attach.go [hunk 2]
```diff
@@ -87,10 +87,19 @@ Example - Attach file to the manifest tagged 'v1' in an OCI image layout folder
 		PreRunE: func(cmd *cobra.Command, args []string) error {
 			opts.RawReference = args[0]
 			opts.FileRefs = args[1:]
-			if err := option.Parse(cmd, &opts); err != nil {
-				return err
+			err := option.Parse(cmd, &opts)
+			if err == nil && opts.Reference == "" {
+				// ensure reference is not empty
+				err = oerrors.NewErrEmptyTagOrDigest(opts.RawReference, cmd, true)
 			}
```
Comment (shizhMSFT) [inline cmd/oras/root/attach.go:94]:
```suggestion
			if err == nil {
				// ensure reference is not empty
				if opts.Reference != "" {
					return nil
				}
				err = oerrors.NewErrEmptyTagOrDigest(opts.RawReference, cmd, true)
			}
```

File: cmd/oras/root/attach.go [hunk 3]
```diff
@@ -87,10 +87,19 @@ Example - Attach file to the manifest tagged 'v1' in an OCI image layout folder
...
-			return nil
+			if len(opts.FileRefs) == 0 {
+				// no file argument provided
+				if err, ok := err.(*oerrors.Error); ok && err.OperationType == oerrors.ParseArtifactReference {
+					// invalid reference
+					err.Recommendation = fmt.Sprintf("Have you specified an artifact reference to attach to? %s", err.Recommendation)
```
Comment (shizhMSFT) [inline cmd/oras/root/attach.go:99]: Consider compiler style recommendations: "Are you missing an artifact reference to attach to"?

Reference:
- https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/assembly-references
- https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs1061

**Review by shizhMSFT — CHANGES_REQUESTED** [review]

File: cmd/oras/root/attach.go (hunk 3 above)
Comment (FeynmanZhou) [reply cmd/oras/root/attach.go:99]: For this case, how about improving the error message as follows: 

```
> oras attach --artifact-type test/example $ARTIFACT ./layers 
Error: "./layers" is an invalid artifact reference.
Usage: oras attach [flags] --artifact-type=<type> <name>{:<tag>|@<digest>} <file>[:<layer_media_type>] | --annotation "key=val"
To attach files to an object, please specify a valid artifact reference and files via arguments in sequence". Run "oras attach -h" for more options and examples.
```

**Review by FeynmanZhou — COMMENTED** [review]
Just one suggested error message for this case

File: cmd/oras/root/attach.go (hunk 3 above)
Comment (qweeah) [reply cmd/oras/root/attach.go:99]: @FeynmanZhou What if a user is trying to attach end-of-life annotations and input an invalid artifact reference? E.g. forget to include the subject's tag. Will the error message `please specify a valid artifact reference and files via arguments in sequence` makes he/she wondering what are the `files` for?

File: cmd/oras/root/attach.go [hunk 4]
```diff
@@ -87,10 +87,22 @@ Example - Attach file to the manifest tagged 'v1' in an OCI image layout folder
...
 			opts.FileRefs = args[1:]
-			if err := option.Parse(cmd, &opts); err != nil {
-				return err
+			err := option.Parse(cmd, &opts)
+			if err == nil {
+				if opts.Reference != "" {
+					return nil
+				}
+				// ensure reference is not empty
+				err = oerrors.NewErrEmptyTagOrDigest(opts.RawReference, cmd, true)
```
Comment (shizhMSFT) [inline cmd/oras/root/attach.go:96]: nit: I just noticed this block of code is `EnsureReferenceNotEmpty` that you have removed. Consider
```suggestion
				err = opts.EnsureReferenceNotEmpty(cmd, true)
				if err == nil {
					return nil
				}
```
The method name is self-explainable.

**Review by shizhMSFT — APPROVED** [review]
LGTM with suggestions

**Review by FeynmanZhou — APPROVED** [review]
LGTM

**Pull request merged**